src/lib/kernel/SimOS.cc

```diff
@@ -0,0 +1,215 @@
+#include "simeng/kernel/SimOS.hh"
+
+/** The size of each time slice a process has. */
+static constexpr uint64_t execTicks = 30000;
+
+namespace simeng {
+namespace kernel {
+
+SimOS::SimOS(std::string executablePath,
+             std::vector<std::string> executableArgs,
+             std::shared_ptr<simeng::memory::Mem> mem)
+    : executablePath_(executablePath),
+      executableArgs_(executableArgs),
+      memory_(mem),
+      syscallHandler_(std::make_shared<SyscallHandler>(processes_)) {
+  createInitialProcess();
+
+  // Create the Special Files directory if indicated to do so in Config file
+  if (Config::get()["CPU-Info"]["Generate-Special-Dir"].as<bool>() == true)
+    createSpecialFileDirectory();
+}
+
+void SimOS::tick() {
+  // Check if simulation halted
+  if (halted_) {
+    return;
+  }
+
+  if (processes_.size() == 0) {
+    halted_ = true;
+    return;
+  }
+
+  /** Scheduling behaviour :
+   * 1. All processes start in a waiting state, and are placed in the
+   * 'waitingProc' queue.
+   *
+   * 2. All cores are looped over, checking their status. If there is a process
+   * in the waitingProc queue then currently executing cores will be tested to
+   * see if they should perform a context switch (Each Process is given
+   * execTicks cycles at a time; round-robin style).
+   *
+   * 3. If a core is sent an interrupt signal, the head of waitingProcs queue is
+   * put into scheduled queue.
+   *
+   * 4. Idle cores can only be scheduled processes from the scheduledProcs
+   * queue.
+   *
+   * 5. When a process is successfully scheduled, it is removed from the
+   * scheduledProcs queue.
+   *
+   * 6. When a process goes back into a waiting state (post context switch) it
+   * will be pushed to the back of the waitingProcs queue.
+   *
+   * This ensures that multiple cores will not be interrupted for a single
```

Comment:
The use of the waiting and scheduled queue, the usage of which is detailed in the comment above, ensures that once a waiting process has interupted a executing core it is moved to the scheduled queue where it waits for (any core, but most likely) the core it interupted to go into an idle state; thereby only ever interupting one core

Reply:
Would be good if the comment could be more explicit that it's the use of those queues which ensures this. Also, I think it would be beneficial to state in 2 that an interrupt signal is used to test if a core should context switch and in 3, that it's when an interrupt signal is successful. Additionally, in 4, a small rewording to make it fit into the style would be beneficial. E.g. "When scheduling a process from the scheduledProcs queue, only cores in an idle state can take on its processing", something of that nature.